docs/sources/mimir/operators-guide/architecture/query-sharding/index.md

```diff
@@ -139,6 +139,17 @@ parallel by the query-frontend, multiplying the previously set value of
 `-querier.max-query-parallelism` by
 `-query-frontend.query-sharding-total-shards`.
 
+## Cardinality-estimation for query sharding (experimental)
+Since a high number of parallel sharded queries increases load on queriers and their dependencies, it is desirable to shard queries only as much as necessary.
+Queries that return more series (i.e., are of high cardinality) need to fetch more data and should therefore be split into a larger batch of shards, while queries that return few or no series should be executed with fewer or no shards at all.
+When determining the number of shards to use for a given query, the sharding logic can optionally take into account the cardinality observed during previous executions of the same query for similar time ranges.
+
+To enable this experimental feature, set `-query-frontend.query-sharding-max-series-per-shard` to a value representing roughly how many series each shard should fetch, and enable result caching (the result cache is used to store cardinality estimates).
```

Comment:
Sorry I'm late to this feature review, feel free to redirect me to RTFM or PR comments, but shouldn't this flag be `min-series-per-shard`? I understand that we want to have _at least_ a given number of series per shard.

Reply:
Oh okay, maybe it's because I assumed that we start with the default sharding config as an upper bound and we restrict the number of shards, but instead of that this invalidates the default config and sets an unbounded number of shards?

Reply:
The idea is that we shard a query such that each individual shard is expected to fetch _at most_ this number of series, that's how the naming came about.

Reply:
Does this mean that with `-query-frontend.query-sharding-max-series-per-shard=10` and 1M series, we're going to have 100K shards? Or is there any other limit involved?

Reply:
I think the logic here is that you start from a default of no sharding at all, and shard as long as you have to until this condition is satisfied, but no more.

Reply:
No, the previous limits all still hold, this will never lead to more shards than would have been used without this feature.

Reply:
Okay, thanks, that makes it clear. Can you add this sentence to the docs please?